doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,776 @@
+
+```
+VLAN_INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attributes
+key         = VLAN_INTERFACE:ifname          ; VLAN Interface name
+;value annotations
+ifname      = 1*64VCHAR                      ; name of the Interface (VLAN)
+; field     = value
+mpls        = "enable"/"disable"             ; Enable/disable MPLS function. Default "disable"
+```
+
+###### CRM Config
+The existing CRM Config stanza is enhanced to include new MPLS in-segment entry and MPLS next-hop attributes.  These attributes parallel existing CRM configuration for other resource types (eg, IPv4/IPv6 routes and next-hops).
+
+```
+CRM
+  Config
+    "mpls_inseg_threshold_type":{{percentage|used|free}} (OPTIONAL)
+    "mpls_inseg_high_threshold":{{UINT32}} (OPTIONAL)
+    "mpls_inseg_low_threshold":{{UINT32}} (OPTIONAL)
+
+    "mpls_nexthop_threshold_type":{{percentage|used|free}} (OPTIONAL)
+    "mpls_nexthop_high_threshold":{{UINT32}} (OPTIONAL)
+    "mpls_nexthop_low_threshold":{{UINT32}} (OPTIONAL)
+
+; Defines schema for CRM MPLS in-segment entry and MPLS next-hop configuration attributes
+; field                    = value
+mpls_inseg_threshold_type  = "percentage"/"used"/"free" ; Threshold type. Default "percentage"
+mpls_inseg_high_threshold  = UINT32             ; High threshold. Default value = 85
+mpls_inseg_low_threshold   = UINT32             ; Low threshold. Default value = 70
+
+mpls_nexthop_threshold_type  = "percentage"/"used"/"free" ; Threshold type. Default "percentage"
+mpls_nexthop_high_threshold  = UINT32             ; High threshold. Default value = 85
+mpls_nexthop_low_threshold   = UINT32             ; Low threshold. Default value = 70
+```
+
+#### ASIC DB
+
+##### ROUTER INTERFACE
+Support for a new attribute is introduced to the ASIC_DB for the existing ROUTER_INTERFACE object type:
+
+##### INSEG ENTRY
+Support for a new object type INSEG_ENTRY is introduced to the ASIC_DB:
+
+##### NEXT HOP
+Support for new attributes are introduced to the ASIC_DB for the existing NEXT_HOP object type:
+
+#### Software Modules
+This section describes modifications to SONiC infrastructure software modules to support MPLS.
+
+##### NetLink Library
+The Netlink library (libnl3) is an existing open source library imported by SONiC to parse and format Netlink messages.
+Modifications to the existing NetLink library MPLS implementation were needed to support MPLS attributes.
+###### Functions
+New Netlink message next-hop accessors were added to retrieve attributes in a nested MPLS encapsulation (RTA_ENCAP_TYPE of LWTUNNEL_IPTUNNEL_MPLS) stanza.  The following accessors retrieve the value for the attributes of MPLS next-hop destination (MPLS_IPTUNNEL_DST) and TTL (MPLS_IPTUNNEL_TTL):
+```
+  /* Accessor to retrieve MPLS destination */
+  extern struct nl_addr *	rtnl_route_nh_get_encap_mpls_dst(struct rtnl_nexthop *);
```

Comment:
>rtnl_route_nh_get_encap_mpls_dst [](http://example.com/codeflow?start=26&length=32)

It's not a good idea to add new functions to upstream libraries. Since it is only used in SONiC applications, why not defined in a repo which belong to SONiC?

Reply:
@qiluo-msft It is not possible to write these accessor functions outside of the context of libnl since the required header files 
 with structure definitions are private to libnl.

Reply:
Could you spend extra effort to submit this patch to upstream repo? that will make long term maintenance easier.

Reply:
@qiluo-msft As I have mentioned in several review meetings, I have already created a PR to upstream these changes to the original libnl project, but the libnl project does not appear to be particularly active and there has not been much activity in December 2020.  https://github.com/thom311/libnl/pull/284